archive/imo/imo2019_q2.lean

```diff
@@ -0,0 +1,589 @@
+/-
+Copyright (c) 2022 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Myers
+-/
+import geometry.euclidean.angle.sphere
+
+/-!
+# IMO 2019 Q2
+
+In triangle `ABC`, point `A₁` lies on side `BC` and point `B₁` lies on side `AC`. Let `P` and
+`Q` be points on segments `AA₁` and `BB₁`, respectively, such that `PQ` is parallel to `AB`.
+Let `P₁` be a point on line `PB₁`, such that `B₁` lies strictly between `P` and `P₁`, and
+`∠PP₁C = ∠BAC`. Similarly, let `Q₁` be a point on line `QA₁`, such that `A₁` lies strictly
+between `Q` and `Q₁`, and `∠CQ₁Q = ∠CBA`.
+
+Prove that points `P`, `Q`, `P₁`, and `Q₁` are concyclic.
+
+We apply the following conventions for formalizing IMO geometry problems. A problem is assumed
+to take place in the plane unless that is clearly not intended, so it is not required to prove
+that the points are coplanar (whether or not that in fact follows from the other conditions).
+Angles in problem statements are taken to be unoriented. A reference to an angle `∠XYZ` is taken
+to imply that `X` and `Z` are not equal to `Y`, since choices of junk values play no role in
+informal mathematics, and those implications are included as hypotheses for the problem whether
+or not they follow from the other hypotheses. Similar, a reference to `XY` as a line is taken to
+imply that `X` does not equal `Y` and that is included as a hypothesis, and a reference to `XY`
+being parallel to something is considered a reference to it as a line. However, such an implicit
+hypothesis about two points being different is included only once for any given two points (even
+if it follows from more than one reference to a line or an angle), if `X ≠ Y` is included then
+`Y ≠ X` is not included separately, and such hypotheses are not included in the case where there
+is also a reference in the problem to a triangle including those two points, or to strict
+betweenness of three points including those two. If betweenness is stated, it is taken to be
+strict betweenness. However, segments and sides are taken to include their endpoints (unless
+this makes a problem false), although those degenerate cases might not necessarily have been
+considered when the problem was formulated and contestants might not have been expected to deal
+with them. A reference to a point being on a side or a segment is expressed directly with `wbtw`
+rather than more literally with `affine_segment`.
```

Comment:
Could you make this a library note? It certainly is more applicable than this single problem (although I know getting a second one ready will take ages 😬)

Reply:
Now marked as a library note.